addons/hw_drivers/drivers/SerialBaseDriver.py

```diff
@@ -0,0 +1,175 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import logging
+import time
+import traceback
+from abc import abstractmethod
+from collections import namedtuple
+from threading import Lock
+from contextlib import contextmanager
+
+import serial
+
+from odoo.tools.translate import _
+from odoo.addons.hw_drivers.controllers.driver import Driver, event_manager
+
+_logger = logging.getLogger(__name__)
+
+SerialProtocol = namedtuple(
+    'SerialProtocol',
+    "name baudrate bytesize stopbits parity timeout writeTimeout measureRegexp statusRegexp "
+    "commandTerminator commandDelay measureDelay newMeasureDelay disable "
+    "measureCommand emptyAnswerValid autoResetWeight")
+
+
+@contextmanager
+def serial_connection(path, protocol, is_probing=False):
+    """Opens a serial connection to a device and closes it automatically after use.
+
+    :param path: path to the device
+    :type path: string
+    :param protocol: an object containing the serial protocol to connect to a device
+    :type protocol: namedtuple
+    :param is_probing: a flag thet if set to `True` makes the timeouts longer, defaults to False
+    :type is_probing: bool, optional
+    """
+
+    PROBING_TIMEOUT = 1
+    port_config = {
+        'baudrate': protocol.baudrate,
+        'bytesize': protocol.bytesize,
+        'stopbits': protocol.stopbits,
+        'parity': protocol.parity,
+        'timeout': PROBING_TIMEOUT if is_probing else protocol.timeout,               # longer timeouts for probing
+        'writeTimeout': PROBING_TIMEOUT if is_probing else protocol.writeTimeout      # longer timeouts for probing
+    }
+    connection = serial.Serial(path, **port_config)
+    yield connection
+    connection.close()
+
+
+class DeviceStatus(object):
+    """Represents the current status of a device."""
```

Comment:
is this new class supposed to be used outside of the serial drivers someday?

Reply:
I doubt it. It seemed clearer somehow to have it separate, but I can integrate the methods in the base serial driver if that's what you mean.